File: src/main/java/io/github/vehkiya/service/DataProvider.java
package io.github.vehkiya.service;

import io.github.vehkiya.data.model.domain.Item;
import org.apache.commons.lang3.StringUtils;

import java.io.IOException;
import java.util.Map;
import java.util.Optional;

public interface DataProvider {

    Map<String, Item> itemsCache();

    void refresh() throws IOException;

    default Optional<Item> findByName(String itemName) {
        return Optional.ofNullable(itemName)
                .map(name -> StringUtils.normalizeSpace(name.trim()))
                .map(name -> itemsCache().get(name));
    }
}

File: src/main/java/io/github/vehkiya/service/DataProvider.kt
package io.github.vehkiya.service

import io.github.vehkiya.data.model.domain.Item
import org.apache.commons.lang3.StringUtils

interface DataProvider {

    fun itemsCache(): Map<String, Item>

    fun refresh()

    fun findByName(itemName: String): Item? {
        val normalizedName = StringUtils.normalizeSpace(itemName.trim())
        return itemsCache()[normalizedName]
    }
}